src/Microsoft.AspNet.Mvc.Razor/Precompilation/PrecompilationTagHelperTypeResolver.cs

```diff
@@ -0,0 +1,138 @@
+            {
+                if (_openGenericDictionaryTypeSymbol == null)
+                {
+                    _openGenericDictionaryTypeSymbol = _compilation.GetTypeByMetadataName(
+                        typeof(IDictionary<,>).FullName);
+
+                    Debug.Assert(_openGenericDictionaryTypeSymbol != null);
+                }
+
+                return _openGenericDictionaryTypeSymbol;
+            }
+        }
+
+        // Internal for unit testing
+        internal IEnumerable<ITypeInfo> GetExportedTypes(string assemblyName)
+        {
+            if (string.Equals(_compilation.AssemblyName, assemblyName, StringComparison.Ordinal))
+            {
+                return GetExportedTypes(_compilation.Assembly);
+            }
+            else
+            {
+                foreach (var reference in _compilation.References)
+                {
+                    var compilationReference = reference as CompilationReference;
+                    if (compilationReference != null &&
+                        string.Equals(
+                            compilationReference.Compilation.AssemblyName,
+                            assemblyName,
+                            StringComparison.Ordinal))
+                    {
+                        return GetExportedTypes(compilationReference.Compilation.Assembly);
+                    }
+
+                    var assemblyReference = reference as PortableExecutableReference;
+                    if (assemblyReference != null)
+                    {
+                        var assemblyOrModuleSymbol = _compilation.GetAssemblyOrModuleSymbol(reference);
+                        if (assemblyOrModuleSymbol != null &&
+                            assemblyOrModuleSymbol.Kind == SymbolKind.Assembly)
+                        {
+                            var assemblySymbol = (IAssemblySymbol)assemblyOrModuleSymbol;
+                            if (string.Equals(
+                                    assemblySymbol.Identity.Name,
+                                    assemblyName,
+                                    StringComparison.Ordinal))
+                            {
+                                return GetExportedTypes(assemblySymbol);
+                            }
+                        }
+                    }
+                }
+
+                return Enumerable.Empty<ITypeInfo>();
+            }
+        }
+
+        private List<ITypeInfo> GetExportedTypes(IAssemblySymbol assembly)
+        {
+            var exportedTypes = new List<ITypeInfo>();
+            GetExportedTypes(assembly.GlobalNamespace, exportedTypes);
+            return exportedTypes;
+        }
+
+        private void GetExportedTypes(INamespaceSymbol namespaceSymbol, List<ITypeInfo> exportedTypes)
+        {
+            foreach (var type in namespaceSymbol.GetTypeMembers().Where(IsPublicType))
+            {
+                exportedTypes.Add(new SymbolBasedTypeInfo(type, OpenGenericDictionaryType));
```

Comment:
Should refactor this out. Shouldn't need to pas `OpenGenericDictionaryType` for every `SymbolBasedTypeInfo`s.